### PR TITLE
Generation of Dates

### PR DESCRIPTION
**Pull request by zinki97**
See https://github.com/jlink/jqwik/issues/140

### CODE REVIEW

File: time/src/main/java/net/jqwik/time/DefaultDateArbitrary.java [hunk 1]
```diff
@@ -70,6 +71,13 @@ public DateArbitrary monthLessOrEqual(Month max) {
 		return clone;
 	}
 
+	@Override
+	public DateArbitrary onlyMonths(Month... months) {
+		DefaultDateArbitrary clone = typedClone();
+		clone.allowedMonths = months;
+		return clone;
+	}
```
Comment (zinki97) [inline time/src/main/java/net/jqwik/time/DefaultDateArbitrary.java:79]: How should I handle the value null?

File: time/src/test/java/net/jqwik/api/time/MonthTests.java [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+	@Group
+	class CheckMonthMethods {
+
+		@Property
+		void atTheEarliest(@ForAll("months") Month month) {
+
+			Arbitrary<Month> months = Dates.months().atTheEarliest(month);
+
+			assertAllGenerated(months.generator(1000), m -> {
```
Comment (jlink) [inline time/src/test/java/net/jqwik/api/time/MonthTests.java:33]: Use net.jqwik.api.testing.TestingSupport.assertAllGenerated(...)

Get a random throuch `@ForAll Random random`.

File: time/src/main/java/net/jqwik/time/DefaultDateArbitrary.java (hunk 1 above)
Comment (jlink) [reply time/src/main/java/net/jqwik/time/DefaultDateArbitrary.java:97]: You mean if months is null?

That's not legal. But I don't think this is checked elsewhere.

File: time/src/main/java/net/jqwik/time/DefaultDateArbitrary.java [hunk 3]
```diff
@@ -8,9 +8,147 @@
...
 	@Override
 	protected Arbitrary<LocalDate> arbitrary() {
-		return Arbitraries.just(LocalDate.of(2020, 12, 9));
+		optimizeRuntime();
+		Arbitrary<Year> year = generateYears();
+		Arbitrary<Month> month = generateMonths();
+		Arbitrary<Integer> dayOfMonth = generateDayOfMonths();
+		return Combinators.combine(year, month, dayOfMonth)
+						  .as(this::generateValidDateFromValues)
+						  .filter(v -> v != null && !v.isBefore(dateMin) && !v.isAfter(dateMax) && isInAllowedDayOfWeeks(v.getDayOfWeek()));
```
Comment (jlink) [inline time/src/main/java/net/jqwik/time/DefaultDateArbitrary.java:30]: Heavy filtering is usually bad for exhaustive generation, edge case generation and shrinking. I wonder if tuning year, month and dayOfMonth generation could get rid of some of the filter parts.

File: time/src/main/java/net/jqwik/time/DefaultMonthDayArbitrary.java [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+	protected Arbitrary<MonthDay> arbitrary() {
+		DateArbitrary dates = Dates.dates()
+								   .atTheEarliest(LocalDate.of(0, monthDayMin.getMonth(), monthDayMin.getDayOfMonth()))
+								   .atTheLatest(LocalDate.of(0, monthDayMax.getMonth(), monthDayMax.getDayOfMonth()))
+								   .yearBetween(0, 0)
+								   .monthBetween(monthMin, monthMax)
+								   .onlyMonths(allowedMonths)
+								   .dayOfMonthBetween(dayOfMonthMin, dayOfMonthMax);
+		return dates.map(v -> MonthDay.of(v.getMonth(), v.getDayOfMonth()));
+	}
```
Comment (jlink) [inline time/src/main/java/net/jqwik/time/DefaultMonthDayArbitrary.java:30]: this looks very complicated for "only" creating a valid date within a year.
Can year 0 be a lap year? Should Feb 29 be possible.

File: time/src/test/java/net/jqwik/api/time/DaysOfMonthTests.java [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+import java.util.*;
+
+import net.jqwik.api.*;
+
+import static org.assertj.core.api.Assertions.*;
+
+import static net.jqwik.api.testing.TestingSupport.*;
+
+@Group
+public class DaysOfMonthTests {
```
Comment (jlink) [inline time/src/test/java/net/jqwik/api/time/DaysOfMonthTests.java:12]: test classes need not be public

**Review by jlink — CHANGES_REQUESTED** [review]
Some general remarks:
- Please adapt your local code style to jqwik standard and reformat.
- Please add tests for shrinking
- Please add tests for edge cases
- Please add tests for exhaustive generation

File: time/src/main/java/net/jqwik/time/DefaultMonthDayArbitrary.java (hunk 4 above)
Comment (zinki97) [reply time/src/main/java/net/jqwik/time/DefaultMonthDayArbitrary.java:30]: Yes, year 0 can be a lap year. Should I change it? I thought it is good to use existing code instead of writing a completely new arbitrary.

File: time/src/main/java/net/jqwik/api/time/DateArbitrary.java [hunk 6]
```diff
@@ -5,4 +5,51 @@
 import net.jqwik.api.*;
 
 public interface DateArbitrary extends Arbitrary<LocalDate> {
+
+	//TODO: Documentation
```
[the remark this answers is not in the document]
Comment (zinki97) [reply time/src/main/java/net/jqwik/api/time/DateArbitrary.java:9]: yeah I know, I'll add it later

File: time/src/main/java/net/jqwik/time/DefaultDateArbitrary.java (hunk 3 above)
Comment (zinki97) [reply time/src/main/java/net/jqwik/time/DefaultDateArbitrary.java]: The `optimizeRuntime` method makes it possible to filter as few dates as possible.

I put the other methods in `generateDateFromValues`. Is it better now?

File: time/src/main/java/net/jqwik/time/DefaultMonthDayArbitrary.java (hunk 4 above)
Comment (jlink) [reply time/src/main/java/net/jqwik/time/DefaultMonthDayArbitrary.java:29]: > Yes, year 0 can be a lap year. Should I change it? I thought it is good to use existing code instead of writing a completely new arbitrary.

Reuse is good... if it doesn't complicate things. Generating dates and stripping away the year will lead to bad performance concerning shrinking and exhaustive genration and lead to duplicate edge cases.

File: time/src/main/java/net/jqwik/time/DefaultDateArbitrary.java [hunk 7]
```diff
@@ -8,9 +8,147 @@
...
+
+	private Arbitrary<Month> generateMonths(){
+		return Dates.months().between(monthMin, monthMax).only(allowedMonths);
+	}
+
+	private Arbitrary<Integer> generateDayOfMonths(){
+		return Dates.daysOfMonth().between(dayOfMonthMin, dayOfMonthMax);
+	}
+
+	private LocalDate generateValidDateFromValues(Year y, Month m, int d){
```
Comment (jlink) [inline time/src/main/java/net/jqwik/time/DefaultDateArbitrary.java]: Can probably be replaced with a call to `Arbitrary.ignoreException(..)`.

File: time/src/main/java/net/jqwik/time/DefaultMonthDayArbitrary.java (hunk 4 above)
Comment (zinki97) [reply time/src/main/java/net/jqwik/time/DefaultMonthDayArbitrary.java:29]: even if the year is always 0?

File: time/src/main/java/net/jqwik/time/DefaultDateArbitrary.java (hunk 3 above)
Comment (jlink) [reply time/src/main/java/net/jqwik/time/DefaultDateArbitrary.java]: Maybe `optimizeRuntime` is just a name I don't understand. It looks to me like some of the things therein should either be done when setting the values or be checked when setting the values for inconsistency - like minYear being greater than maxYear.

File: time/src/main/java/net/jqwik/api/time/DaysOfWeekArbitrary.java [hunk 8]
```diff
@@ -0,0 +1,12 @@
+package net.jqwik.api.time;
+
+import java.time.*;
+
+import net.jqwik.api.*;
+
+public interface DaysOfWeekArbitrary extends Arbitrary<DayOfWeek> {
```
Comment (jlink) [inline time/src/main/java/net/jqwik/api/time/DaysOfWeekArbitrary.java:7]: Since DayOfWeek is just an enum the whole arbitrary class is not necessary or is it. Could be replaced with Arbitraries.of(DayOfWeek.class) I think. only(..) doe not pull its weight either since Arbitraries.of(MONDAY, WEDNESDAY) is just as short.

File: time/src/main/java/net/jqwik/time/DefaultMonthDayArbitrary.java (hunk 4 above)
Comment (jlink) [reply time/src/main/java/net/jqwik/time/DefaultMonthDayArbitrary.java:29]: Not sure. Have you done an experiment to check?

File: time/src/main/java/net/jqwik/api/time/DaysOfWeekArbitrary.java (hunk 8 above)
Comment (zinki97) [reply time/src/main/java/net/jqwik/api/time/DaysOfWeekArbitrary.java:7]: Okay, I'll remove it

File: time/src/main/java/net/jqwik/time/DefaultMonthDayArbitrary.java (hunk 4 above)
Comment (zinki97) [reply time/src/main/java/net/jqwik/time/DefaultMonthDayArbitrary.java:29]: No, but `Dates.years().between(yearMin, yearMax)` creates only one year wit the value 0. Only `generateMonths` and `generateDayOfMonths` have more values. So the year value is always the same and can't change.

**Comment by zinki97** [comment]
Can you please provide feedback on the `ExhaustiveGeneration` classes?

File: time/src/test/java/net/jqwik/api/time/DatesTests.java [hunk 9]
```diff
@@ -1,20 +1,342 @@
...
+			assertThat(value).isEqualTo(LocalDate.of(0, Month.JANUARY, 1));
+		}
+
+		@Property
+		void shrinksToSmallestFailingPositiveValue(@ForAll Random random){
+			DateArbitrary dates = Dates.dates();
+			TestingFalsifier<LocalDate> falsifier = date -> date.isBefore(LocalDate.of(2013, Month.MAY, 25));
+			LocalDate value = shrinkToMinimal(dates, random, falsifier);
+			assertThat(value).isEqualTo(LocalDate.of(2013, Month.MAY, 25));
+		}
```
Comment (zinki97) [inline time/src/test/java/net/jqwik/api/time/DatesTests.java:278]: How can I fix this?

`Expecting:
     <2014-01-01 (java.time.LocalDate)>
    to be equal to:
     <2013-05-25 (java.time.LocalDate)>
    but was not.`

Or should I accept that it can't shrink perfectly?

Comment (jlink) [reply time/src/test/java/net/jqwik/api/time/DatesTests.java:278]: I'd aim for perfect shrinking here. To see why it won't happen requires detailed analysis. Leave it as is, and create a list with things that don't work as expected yet.

**Comment by jlink** [comment]
> Can you please provide feedback on the `ExhaustiveGeneration` classes?

Tests for exhaustive generation look fine to me. 

I'd probably also look at a full generation for all MonthAndDay values. But I may have missed that.

File: time/src/test/java/net/jqwik/api/time/DatesTests.java (hunk 9 above)
Comment (zinki97) [reply time/src/test/java/net/jqwik/api/time/DatesTests.java:278]: List is here: https://github.com/jlink/jqwik/issues/140#issuecomment-743322513

File: time/src/test/java/net/jqwik/api/time/DatesTests.java [hunk 10]
```diff
@@ -1,20 +1,384 @@
...
+																																																															 .of(2020, Month.DECEMBER, 21), LocalDate
+																																																																									.of(2020, Month.DECEMBER, 24), LocalDate
+																																																																																		   .of(2020, Month.DECEMBER, 28), LocalDate
+																																																																																												  .of(2020, Month.DECEMBER, 31));
+		}
+
+	}
+
+	@Group
+	class EdgeCasesTests {
```
Comment (zinki97) [inline time/src/test/java/net/jqwik/api/time/DatesTests.java:343]: Can you please provide feedback on the `EdgeCasesTests` classes?

Some of the tests are currently not working.
For example:
```
@Property(tries = 5)
void between() {

	MonthArbitrary months = Dates.months().between(Month.MARCH, Month.AUGUST);
	Set<Month> edgeCases = collectEdgeCases(months.edgeCases());
	assertThat(edgeCases).hasSize(2);
	assertThat(edgeCases).contains(Month.MARCH, Month.AUGUST);

}
```

Comment (jlink) [reply time/src/test/java/net/jqwik/api/time/DatesTests.java:343]: What's the error?

Comment (zinki97) [reply time/src/test/java/net/jqwik/api/time/DatesTests.java:343]:
```
java.lang.AssertionError:
    Expected size:<2> but was:<0> in:
    <[]>
```

Comment (jlink) [reply time/src/test/java/net/jqwik/api/time/DatesTests.java:343]: I'll have a closer look. 
BTW, `@Property(tries = 5)` doesnt make sense IMO since edge case generation is deterministic.
I'd go with `@Example`

Comment (jlink) [reply time/src/test/java/net/jqwik/api/time/DatesTests.java:343]: If you take away the "smallest" and "largest" month, no edge case remains which is what I'd expect. So I think you're good here.

Comment (jlink) [reply time/src/test/java/net/jqwik/api/time/DatesTests.java:343]: Similiar to edge cases, exhaustive generation is also deterministic and thus I'd go with `@Example` there.

File: time/src/test/java/net/jqwik/api/time/MonthTests.java [hunk 11]
```diff
@@ -0,0 +1,204 @@
...
+			MonthArbitrary months = Dates.months().between(Month.MARCH, Month.AUGUST);
+			Set<Month> edgeCases = collectEdgeCases(months.edgeCases());
+			assertThat(edgeCases).hasSize(2);
+			assertThat(edgeCases).contains(Month.MARCH, Month.AUGUST);
+
+		}
+
+	}
+
+	public static Arbitrary<Month[]> generateMonths() {
```
Comment (jlink) [inline time/src/test/java/net/jqwik/api/time/MonthTests.java:181]: Maybe I dont undestand it, but if it's supposed to generate an Array of Month, I'd do it with `monthArbitrary.array(Month[].class).ofSize...`

File: time/src/test/java/net/jqwik/api/time/DatesTests.java (hunk 10 above)
Comment (zinki97) [reply time/src/test/java/net/jqwik/api/time/DatesTests.java:343]: `months = months.edgeCases(monthConfig -> monthConfig.add(min).add(max));`

I fixed it with that and I am working on the other edge cases. Should I remove it?

Comment (zinki97) [reply time/src/test/java/net/jqwik/api/time/DatesTests.java:343]: As user I expect the min and max values as edgecases

Comment (jlink) [reply time/src/test/java/net/jqwik/api/time/DatesTests.java:343]: Using `EdgeCases.Config.add(..)` has the drawback that values added this way cannot be shrunk. 
Given there are only 12 months anyway, I'd drop it here.

Comment (jlink) [reply time/src/test/java/net/jqwik/api/time/DatesTests.java:343]: > As user I expect the min and max values as edgecases

You could manipulate the original selection to go from min to max. Then those edge cases would get in automatically.

Comment (zinki97) [reply time/src/test/java/net/jqwik/api/time/DatesTests.java:343]: > Using `EdgeCases.Config.add(..)` has the drawback that values added this way cannot be shrunk.
> Given there are only 12 months anyway, I'd drop it here.

Sure that this is a problem? The shrinking tests are not bader now.

Comment (zinki97) [reply time/src/test/java/net/jqwik/api/time/DatesTests.java:343]: I'll upload it so you can take a closer look.

File: time/src/test/java/net/jqwik/api/time/MonthTests.java (hunk 11 above)
Comment (zinki97) [reply time/src/test/java/net/jqwik/api/time/MonthTests.java:181]: The arbitrary generates arrays with different months and lengths

Comment (jlink) [reply time/src/test/java/net/jqwik/api/time/MonthTests.java:181]: So array() is the right thing I guess.

File: time/src/main/java/net/jqwik/time/DefaultMonthArbitrary.java [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+import net.jqwik.api.time.*;
+
+public class DefaultMonthArbitrary extends ArbitraryDecorator<Month> implements MonthArbitrary {
+
+	private Month min = Month.JANUARY;
+	private Month max = Month.DECEMBER;
+	private Month[] allowedMonths = new Month[]{Month.JANUARY, Month.FEBRUARY, Month.MARCH, Month.APRIL, Month.MAY, Month.JUNE, Month.JULY, Month.AUGUST, Month.SEPTEMBER, Month.OCTOBER, Month.NOVEMBER, Month.DECEMBER};
+
+	@Override
+	protected Arbitrary<Month> arbitrary() {
```
Comment (jlink) [inline time/src/main/java/net/jqwik/time/DefaultMonthArbitrary.java:16]: I suggest the following implementation

	@Override
	protected Arbitrary<Month> arbitrary() {
		List<Month> values = Arrays.asList(Month.class.getEnumConstants());
		final int indexMin = values.indexOf(min);
		final int indexMax = values.indexOf(max);

		return Arbitraries.integers()
						  .between(indexMin, indexMax)
						  .edgeCases(integerConfig -> integerConfig.includeOnly(indexMin, indexMax))
						  .map(values::get)
						  .filter(this::isInAllowedMonths);
	}

This will break one exhaustive generation test because this generation is more precise.

File: time/src/test/java/net/jqwik/api/time/MonthDayTests.java [hunk 13]
```diff
@@ -0,0 +1,333 @@
...
+
+			}
+
+			@Property
+			void monthOnlyMonths(@ForAll("monthsOnlyMonths") Month[] months, @ForAll Random random) {
+
+				Arbitrary<MonthDay> dates = Dates.monthDays().onlyMonths(months);
+
+				assertAllGenerated(dates.generator(1000), random, md -> {
+					assertThat(md.getMonth()).isIn(months);
```
Comment (jlink) [inline time/src/test/java/net/jqwik/api/time/MonthDayTests.java:144]: This generates a vararg warning. Use List<Month> instead of Month[]. This happens in a few other places as well.

File: time/src/test/java/net/jqwik/api/time/MonthTests.java (hunk 11 above)
Comment (jlink) [reply time/src/test/java/net/jqwik/api/time/MonthTests.java:181]: This is my suggestion:

	public static Arbitrary<Month[]> generateMonths() {
		Arbitrary<Month> monthArbitrary = Arbitraries.of(Month.class);
		return monthArbitrary.array(Month[].class).ofMinSize(1);
	}

Better even, use a `Set<Month>` but then you don't need the whole method anymore becaus you can just say ` @ForAll @Size(min=1) Set<Month>` at point of usage.

File: time/src/main/java/net/jqwik/time/DefaultMonthArbitrary.java (hunk 12 above)
Comment (jlink) [reply time/src/main/java/net/jqwik/time/DefaultMonthArbitrary.java:16]: It can even be further simplified if we drop `MonthArbitrary.includeOnly(..)` which is just a complicated way of saying `Arbitrary.of(..)`.

File: time/src/test/java/net/jqwik/api/time/MonthDayTests.java (hunk 13 above)
Comment (zinki97) [reply time/src/test/java/net/jqwik/api/time/MonthDayTests.java:145]: I think the warning has now disappeared by using Set <month>.

File: time/src/main/java/net/jqwik/time/DefaultDateArbitrary.java [hunk 14]
```diff
@@ -8,9 +8,135 @@
 
 public class DefaultDateArbitrary extends ArbitraryDecorator<LocalDate> implements DateArbitrary {
```
Comment (jlink) [inline time/src/main/java/net/jqwik/time/DefaultDateArbitrary.java:9]: All Default* classes should get a
`@API(status = INTERNAL)` annotation.

File: time/src/main/java/net/jqwik/api/time/DateArbitrary.java [hunk 15]
```diff
@@ -2,7 +2,72 @@
...
+
+	/**
+	 * Set the allowed lower {@code min} (included) and upper {@code max} (included) bounder of generated year values.
+	 */
+	DateArbitrary yearBetween(Year min, Year max);
+
+	/**
+	 * Set the allowed lower {@code min} (included) and upper {@code max} (included) bounder of generated year values.
+	 */
+	default DateArbitrary yearBetween(int min, int max) {
```
Comment (jlink) [inline time/src/main/java/net/jqwik/api/time/DateArbitrary.java:42]: Min/max int values should be documented.

File: time/src/main/java/net/jqwik/api/time/DateArbitrary.java [hunk 16]
```diff
@@ -2,7 +2,72 @@
...
+
+	/**
+	 * Set the allowed lower {@code min} (included) and upper {@code max} (included) bounder of generated month values.
+	 */
+	DateArbitrary monthBetween(Month min, Month max);
+
+	/**
+	 * Set the allowed lower {@code min} (included) and upper {@code max} (included) bounder of generated month values.
+	 */
+	default DateArbitrary monthBetween(int min, int max) {
```
Comment (jlink) [inline time/src/main/java/net/jqwik/api/time/DateArbitrary.java:54]: Min/max int values should be documented.

File: time/src/main/java/net/jqwik/api/time/DateArbitrary.java [hunk 17]
```diff
@@ -2,7 +2,72 @@
...
+
+	/**
+	 * Set an array of allowed {@code months}.
+	 */
+	DateArbitrary onlyMonths(Month... months);
+
+	/**
+	 * Set the allowed lower {@code min} (included) and upper {@code max} (included) bounder of generated day of month values.
+	 */
+	DateArbitrary dayOfMonthBetween(int min, int max);
```
Comment (jlink) [inline time/src/main/java/net/jqwik/api/time/DateArbitrary.java:66]: Min/max int values should be documented.

File: time/src/main/java/net/jqwik/api/time/DaysOfMonthArbitrary.java [hunk 18]
```diff
@@ -0,0 +1,32 @@
...
+/**
+ * Fluent interface to configure the generation of day of month values.
+ */
+@API(status = EXPERIMENTAL, since = "1.4.0")
+public interface DaysOfMonthArbitrary extends Arbitrary<Integer> {
+
+	/**
+	 * Set the allowed lower {@code min} (included) and upper {@code max} (included) bounder of generated day of month values.
+	 */
+	default DaysOfMonthArbitrary between(int min, int max) {
```
Comment (jlink) [inline time/src/main/java/net/jqwik/api/time/DaysOfMonthArbitrary.java:18]: Min/max int values should be documented.

File: time/src/main/java/net/jqwik/api/time/DaysOfMonthArbitrary.java [hunk 19]
```diff
@@ -0,0 +1,32 @@
...
+		return greaterOrEqual(min).lessOrEqual(max);
+	}
+
+	/**
+	 * Set the allowed lower {@code min} (included) bounder of generated day of month values.
+	 */
+	DaysOfMonthArbitrary greaterOrEqual(int min);
```
Comment (jlink) [inline time/src/main/java/net/jqwik/api/time/DaysOfMonthArbitrary.java:25]: Min/max int values should be documented.

File: time/src/main/java/net/jqwik/api/time/DaysOfMonthArbitrary.java [hunk 20]
```diff
@@ -0,0 +1,32 @@
...
+
+	/**
+	 * Set the allowed upper {@code max} (included) bounder of generated day of month values.
+	 */
+	DaysOfMonthArbitrary lessOrEqual(int max);
```
Comment (jlink) [inline time/src/main/java/net/jqwik/api/time/DaysOfMonthArbitrary.java:30]: Min/max int values should be documented.

File: time/src/main/java/net/jqwik/api/time/MonthArbitrary.java [hunk 21]
```diff
@@ -0,0 +1,60 @@
...
+	 * Set the allowed lower {@code min} (included) and upper {@code max} (included) bounder of generated month values.
+	 */
+	default MonthArbitrary between(Month min, Month max) {
+		return atTheEarliest(min).atTheLatest(max);
+	}
+
+	/**
+	 * Set the allowed lower {@code min} (included) and upper {@code max} (included) bounder of generated month values.
+	 */
+	default MonthArbitrary between(int min, int max) {
```
Comment (jlink) [inline time/src/main/java/net/jqwik/api/time/MonthArbitrary.java:27]: Min/max int values should be documented.

File: time/src/main/java/net/jqwik/api/time/MonthArbitrary.java [hunk 22]
```diff
@@ -0,0 +1,60 @@
...
+
+	/**
+	 * Set the allowed upper {@code max} (included) bounder of generated month values.
+	 */
+	MonthArbitrary atTheLatest(Month max);
+
+	/**
+	 * Set the allowed upper {@code max} (included) bounder of generated month values.
+	 */
+	default MonthArbitrary atTheLatest(int max) {
```
Comment (jlink) [inline time/src/main/java/net/jqwik/api/time/MonthArbitrary.java:51]: Min/max int values should be documented.

File: time/src/main/java/net/jqwik/api/time/MonthDayArbitrary.java [hunk 23]
```diff
@@ -0,0 +1,94 @@
...
+	 * Set the allowed lower {@code min} (included) and upper {@code max} (included) bounder of generated month values.
+	 */
+	default MonthDayArbitrary monthBetween(Month min, Month max) {
+		return monthGreaterOrEqual(min).monthLessOrEqual(max);
+	}
+
+	/**
+	 * Set the allowed lower {@code min} (included) and upper {@code max} (included) bounder of generated month values.
+	 */
+	default MonthDayArbitrary monthBetween(int min, int max) {
```
Comment (jlink) [inline time/src/main/java/net/jqwik/api/time/MonthDayArbitrary.java:44]: Min/max int values should be documented.

File: time/src/main/java/net/jqwik/api/time/MonthDayArbitrary.java [hunk 24]
```diff
@@ -0,0 +1,94 @@
...
+
+	/**
+	 * Set the allowed lower {@code min} (included) bounder of generated month values.
+	 */
+	MonthDayArbitrary monthGreaterOrEqual(Month min);
+
+	/**
+	 * Set the allowed lower {@code min} (included) bounder of generated month values.
+	 */
+	default MonthDayArbitrary monthGreaterOrEqual(int min) {
```
Comment (jlink) [inline time/src/main/java/net/jqwik/api/time/MonthDayArbitrary.java:56]: Min/max int values should be documented.

File: time/src/main/java/net/jqwik/api/time/MonthDayArbitrary.java [hunk 25]
```diff
@@ -0,0 +1,94 @@
...
+	 * Set the allowed lower {@code min} (included) and upper {@code max} (included) bounder of generated day of month values.
+	 */
+	default MonthDayArbitrary dayOfMonthBetween(int min, int max) {
+		return dayOfMonthGreaterOrEqual(min).dayOfMonthLessOrEqual(max);
+	}
+
+	/**
+	 * Set the allowed lower {@code min} (included) bounder of generated day of month values.
+	 */
+	MonthDayArbitrary dayOfMonthGreaterOrEqual(int min);
```
Comment (jlink) [inline time/src/main/java/net/jqwik/api/time/MonthDayArbitrary.java:87]: Min/max int values should be documented.

File: time/src/main/java/net/jqwik/api/time/MonthDayArbitrary.java [hunk 26]
```diff
@@ -0,0 +1,94 @@
...
+
+	/**
+	 * Set the allowed upper {@code max} (included) bounder of generated day of month values.
+	 */
+	MonthDayArbitrary dayOfMonthLessOrEqual(int max);
```
Comment (jlink) [inline time/src/main/java/net/jqwik/api/time/MonthDayArbitrary.java:92]: Min/max int values should be documented.

File: time/src/main/java/net/jqwik/api/time/MonthDayArbitrary.java [hunk 27]
```diff
@@ -0,0 +1,94 @@
...
+
+import net.jqwik.api.*;
+
+import static org.apiguardian.api.API.Status.*;
+
+/**
+ * Fluent interface to configure the generation of month and day values.
+ */
+@API(status = EXPERIMENTAL, since = "1.4.0")
+public interface MonthDayArbitrary extends Arbitrary<MonthDay> {
```
Comment (jlink) [inline time/src/main/java/net/jqwik/api/time/MonthDayArbitrary.java:15]: I'd get rid of all the greate/less/orEqual methods. Too much API maintenance burden for too little value.

File: time/src/main/java/net/jqwik/api/time/YearArbitrary.java [hunk 28]
```diff
@@ -0,0 +1,55 @@
...
+
+import net.jqwik.api.*;
+
+import static org.apiguardian.api.API.Status.*;
+
+/**
+ * Fluent interface to configure the generation of Year values.
+ */
+@API(status = EXPERIMENTAL, since = "1.4.0")
+public interface YearArbitrary extends Arbitrary<Year> {
```
Comment (jlink) [inline time/src/main/java/net/jqwik/api/time/YearArbitrary.java:15]: I'd get rid of all the greate/less/orEqual methods. Too much API maintenance burden for too little value.

File: time/src/main/java/net/jqwik/api/time/YearMonthArbitrary.java [hunk 29]
```diff
@@ -0,0 +1,115 @@
...
+
+import net.jqwik.api.*;
+
+import static org.apiguardian.api.API.Status.*;
+
+/**
+ * Fluent interface to configure the generation of year and month values.
+ */
+@API(status = EXPERIMENTAL, since = "1.4.0")
+public interface YearMonthArbitrary extends Arbitrary<YearMonth> {
```
Comment (jlink) [inline time/src/main/java/net/jqwik/api/time/YearMonthArbitrary.java:15]: I'd get rid of all the greate/less/orEqual methods. Too much API maintenance burden for too little value.

File: documentation/src/docs/user-guide.template.md [hunk 30]
```diff
@@ -1295,6 +1295,55 @@ void restrictedEmailAddresses(@ForAll @Email(quotedLocalPart = false, ipv4Host =
 }
 ```
 
+### Date Arbitrary Types
```
Comment (jlink) [inline documentation/src/docs/user-guide.template.md:1298]: Looking at the other headlines I'd rather write "Generation of Dates".

Using annotations (as soons as there are any) will also be documented here so a more generic headline fits better.

File: documentation/src/docs/user-guide.template.md [hunk 31]
```diff
@@ -1295,6 +1295,55 @@ void restrictedEmailAddresses(@ForAll @Email(quotedLocalPart = false, ipv4Host =
...
+
```
Comment (jlink) [inline documentation/src/docs/user-guide.template.md:1299]: Add a sentence to describe that date generation is in an extra module which must be in a project's dependency.

File: documentation/src/docs/user-guide.template.md [hunk 32]
```diff
@@ -1295,6 +1295,55 @@ void restrictedEmailAddresses(@ForAll @Email(quotedLocalPart = false, ipv4Host =
...
+- [`YearArbitrary years()`](/docs/${docsVersion}/javadoc/net/jqwik/api/Dates.html#years())
+- [`MonthArbitrary months()`](/docs/${docsVersion}/javadoc/net/jqwik/api/Dates.html#months())
+- [`DaysOfMonthArbitrary daysOfMonth()`](/docs/${docsVersion}/javadoc/net/jqwik/api/Dates.html#daysOfMonth())
+- [`YearMonthArbitrary yearMonths()`](/docs/${docsVersion}/javadoc/net/jqwik/api/Dates.html#yearMonths())
+- [`MonthDayArbitrary monthDays()`](/docs/${docsVersion}/javadoc/net/jqwik/api/Dates.html#monthDays())
+
+In addition, you can constrain their values using the following functions:
+
+#### DateArbitrary
+
```
Comment (jlink) [inline documentation/src/docs/user-guide.template.md:1312]: Add a code example or a few for generation of dates. Mention that `LocalDate` is the target type.

**Review by jlink — CHANGES_REQUESTED** [review]

**Review by jlink — CHANGES_REQUESTED** [review]
1. Some test classes no longer compile. I suggest to run clean Gradle build before pushing changes.

2. There are no tests checking statistical distribution. I think they would make sense to check that months, day of week, and day of months are approximately equally distributed.

3. Javadoc of DateArbitrary should mention that all generated dates use the Gregorian Calendar, even if they are before Oct 15, 1582.

4. I'd like to have default min year to be 1900 and max year 2500. So by default only dates between Jan 1, 1900 and Dec 31, 2500 are generated. This must be documented. That change is otional; I can add it myself later.

**Comment by zinki97** [comment]
- [ ] 1.
- [ ] 2.
- [ ] 3.
- [ ] 4.
- [ ] 5.
